src/util/simplify_expr2.cpp

```diff
@@ -1287,9 +1287,51 @@ static expr2tc do_bit_munge_operation(
   // signed or not.
   constant_int2t *theint;
   if(is_signedbv_type(type))
-    theint = new constant_int2t(type, BigInt(val1));
+  {
+    switch(type->get_width())
+    {
+    case 8:
+      theint = new constant_int2t(type, BigInt((int8_t)val1));
+      break;
+    case 16:
+      theint = new constant_int2t(type, BigInt((int16_t)val1));
+      break;
+    case 24:
+      theint = new constant_int2t(type, BigInt((int16_t)val1));
+      break;
+    case 32:
+      theint = new constant_int2t(type, BigInt((int32_t)val1));
+      break;
+    case 64:
+      theint = new constant_int2t(type, BigInt(val1));
+      break;
+    default:
+      abort();
+    }
+  }
   else
-    theint = new constant_int2t(type, BigInt((uint64_t)val1));
+  {
+    switch(type->get_width())
+    {
+    case 8:
+      theint = new constant_int2t(type, BigInt((uint8_t)val1));
+      break;
+    case 16:
+      theint = new constant_int2t(type, BigInt((uint16_t)val1));
+      break;
+    case 24:
+      theint = new constant_int2t(type, BigInt((uint16_t)val1));
+      break;
+    case 32:
+      theint = new constant_int2t(type, BigInt((uint32_t)val1));
+      break;
+    case 64:
+      theint = new constant_int2t(type, BigInt((uint64_t)val1));
+      break;
+    default:
+      abort();
+    }
+  }
```

Comment:
That's a lot of duplicated code... Can we make this code more generic? 

Something like:
```
uint64_t bar = static_cast<uint64_t>(val1) & ((1ULL << type->get_width()) - 1);
return constant_int2tc(type, BigInt(bar));
```

I think it should cover both signed and unsigned cases, otherwise we can keep the `if(is_signedbv_type(type))` branch.

Reply:
I agree with you, @mikhailramalho. 

I have rewritten the code after a quick meeting with @fbrausse. This suggestion seems to be a clean solution; let us see if the CI will not complain about that.